container/openejb-core/src/main/java/org/apache/openejb/junit/jupiter/ApplicationComposerExtension.java

```diff
@@ -0,0 +1,155 @@
+        if (isPerJvm(context) || isPerAll(context)) {
+            doRelease(context);
+        } else if (isPerDefault(context) && isPerClass(context)) {
+            doRelease(context);
+        }
+    }
+
+
+    @Override
+    public void beforeEach(ExtensionContext context) {
+
+        if (isPerEach(context)) {
+            doStart(context);
+        } else if (isPerDefault(context) && !isPerClass(context)) {
+            doStart(context);
+        }
+
+        if (!isPerClass(context)) {
+            doInject(context);
+        }
+    }
+
+    @Override
+    public void afterEach(ExtensionContext context) throws Exception {
+        if (isPerEach(context)) {
+            doRelease(context);
+        } else if (isPerDefault(context) && !isPerClass(context)) {
+            doRelease(context);
+        }
+    }
+
+    private void doRelease(final ExtensionContext extensionContext) throws Exception {
+        if (isPerJvm(extensionContext)) {
+            //FIXME how to release / close after last test class execution?
+            BASE.close();
+        } else {
+            extensionContext.getStore(NAMESPACE).get(ApplicationComposers.class, ApplicationComposers.class).after();
+        }
+    }
+
+    private void doStart(final ExtensionContext extensionContext) {
+
+        Optional<Class<?>> oClazz = extensionContext.getTestClass();
+
+        if (!oClazz.isPresent()) {
+            throw new OpenEJBRuntimeException("Could not get test class from the given extension context.");
+        }
+
+        extensionContext.getStore(NAMESPACE).put(ApplicationComposers.class,
+                new ApplicationComposers(oClazz.get(), getAdditionalModules(oClazz.get())));
+
+    }
+
+    private void doInject(final ExtensionContext extensionContext) {
+        Optional<TestInstances> oTestInstances = extensionContext.getTestInstances();
+
+        if (!oTestInstances.isPresent()) {
+            throw new OpenEJBRuntimeException("No test instances available for the given extension context.");
+        }
+
+        List<Object> testInstances = oTestInstances.get().getAllInstances();
+
+        if (isPerJvm(extensionContext)) {
+            testInstances.forEach(t -> {
+                try {
+                    BASE.composerInject(t);
+                } catch (Exception e) {
+                    throw new OpenEJBRuntimeException(e);
+                }
+            });
+        } else {
+            ApplicationComposers delegate = extensionContext.getStore(NAMESPACE)
+                    .get(ApplicationComposers.class, ApplicationComposers.class);
+
+            testInstances.forEach(t -> {
+                try {
+                    delegate.before(t);
+                } catch (Exception e) {
+                    throw new OpenEJBRuntimeException(e);
+                }
+            });
+        }
+    }
+
+    private Object[] getAdditionalModules(final Class<?> clazz) {
+        return Arrays.stream(clazz.getClasses())
```

Comment:
this likely do not do what is expected and can be random (thinking to @Nested tests of junit5).

ApplicationComposerRule shows the idea: you pass the list of additional modules.

In junit5 it can be:

> @RegisterExtension ApplicationComposerExtension ext = new ApplicationComposerExtension(new MyApp());

it enables to define the app outside the test and/or to compose subapps definitions (like a MyStack() + finish the spec in the test)